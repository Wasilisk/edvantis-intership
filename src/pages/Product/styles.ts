import styled from "styled-components";
import {TabStyles} from "../../components/Tabs/Tab/styles";
import {NumberInputStyles} from "../../components/common/NumberInput/styles";

export const ProductPageStyles = styled.div`
  max-width: 1200px;
  height: 100%;
  margin: 40px auto;
  padding: 0 24px;
  display: flex;


  & > div {
    flex-basis: 50%;
    max-width: 50%;
    padding: 0 15px;
  }

  .product-image {
    display: flex;
    align-items: center;
    justify-content: center;
    margin-bottom: 40px;

    img {
      height: auto;
      width: 90%;
    }
  }

  .product-info {
    .text-info {
      line-height: 1.5;
      letter-spacing: .05em;
      color: #707070;
      position: relative;
      padding-bottom: 48px;

      .not-available {
        color: #d9534c;
      }

      .status {
        margin: 0;
      }

      .product-name {
        font-size: 22px;
        line-height: 1.3;
        font-weight: 500;
        color: #333;
        margin: 0;
      }

      .product-code {
        font-size: 14px;
        line-height: 1.6;
        letter-spacing: .05em;
        color: #858585;
        margin: 8px 0 0 0;
      }

      .price {
        font-weight: 600;
        font-size: 20px;
        color: #333;
        margin: 24px 0 0 0;
      }

      :after {
        content: "";
        position: absolute;
        height: 1px;
        right: 0;
        bottom: 0;
        left: 0;
        background-color: currentColor;
        opacity: .125;
      }
    }

    .actions {
      display: flex;
      align-items: center;
      padding: 24px 0;

      ${NumberInputStyles} {
        margin-right: 20px;
      }
    }

    .additional-info {

      ${TabStyles} {
        padding: 0 10px 10px;

        span {
          text-transform: uppercase;
          cursor: pointer;
          font-weight: 600;
          font-size: 12px;
          line-height: 16px;
        }
      }

      pre {
        white-space: pre-wrap;
        white-space: -moz-pre-wrap;
        white-space: -o-pre-wrap;
        word-wrap: break-word;
        word-break: keep-all;

        font-size: 16px;
        line-height: 1.5;
        font-family: Poppins, sans-serif;
        letter-spacing: .05em;
        color: #707070;
      }
    }
  }

  @media (max-width: 768px) {
    flex-direction: column;

    & > div {
      flex-basis: 100%;
      max-width: 100%;
    }
  }
`;