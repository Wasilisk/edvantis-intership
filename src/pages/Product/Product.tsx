import React, {useEffect, useState} from 'react';
import {useParams} from "react-router-dom";
import {useAppDispatch, useAppSelector} from "../../hooks";
import {getProductBySearchName} from "../../store/slices/Products";
import {selectSelectedProduct} from "../../store/slices/Products/selectors";
import {ProductPageStyles} from './styles';
import Tab from "../../components/Tabs/Tab";
import Tabs from "../../components/Tabs";
import TabPanel from "../../components/Tabs/TabPanel";
import Button from "../../components/common/Button";
import NumberInput from "../../components/common/NumberInput";

const Product = () => {
    const dispatch = useAppDispatch();
    const {productName} = useParams<string>();
    const productInfo = useAppSelector(selectSelectedProduct);
    const [value, setValue] = useState<string | null>(null)
    const [productCount, setProductCount] = useState<number>(1)

    useEffect(() => {
        if (productName) {
            dispatch(getProductBySearchName(productName))
        }
    }, [])

    useEffect(() => {
        if(productInfo) {
            setValue(productInfo.additional_info[0].name)
        }
    }, [productInfo])


    return (
        <ProductPageStyles>
            {
                productInfo && <>
                    <div className="product-image">
                        <img src={productInfo.image.url} alt={productInfo.name}/>
                    </div>
                    <div className="product-info">
                        <div className="text-info">
                            <p
                                className={`status ${productInfo.status === "in stock" ? "highlight" : "not-available"}`}
                            >
                                {productInfo.status.charAt(0).toUpperCase() + productInfo.status.slice(1)}
                            </p>
                            <p className="product-name">{productInfo.name}</p>
                            <p className="product-code">Product code {productInfo.id}</p>
                            <p className="price">{productInfo.price.toFixed(2)}USD</p>
                        </div>
                        <div className="actions">
                            <NumberInput min={1} max={10} value={productCount} onChange={setProductCount}/>
                            <Button variant="filled" size="medium" label="Add to card" fullWidth/>
                        </div>
                        <div className="additional-info">
                            <Tabs>
                                {
                                    productInfo.additional_info.map((tab, index) => {
                                        const clickHandler = () => setValue(tab.name);

                                        return <Tab
                                            key={index}
                                            label={tab.name}
                                            isActive={value === tab.name}
                                            onClick={clickHandler}
                                        />
                                    })
                                }
                            </Tabs>
                            {
                                productInfo.additional_info.map((tab, index) => <TabPanel
                                    key={index}
                                    value={value as string}
                                    index={tab.name}
                                >
                                    <pre>{tab.value}</pre>
                                </TabPanel>)
                            }
                        </div>
                    </div>
                </>
            }
        </ProductPageStyles>
    );
};

export default Product;