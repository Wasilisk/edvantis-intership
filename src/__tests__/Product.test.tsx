import {render} from "./__utils__/store-wrapper";
import {mockProductCard} from "./__mock__/mock-product-card";
import Product from "../pages/Product/Product";

describe('Product component', () => {
    it('should render header component', () => {
        const {asFragment, getByText, getByAltText} = render(<Product/>, {
            initialState: {
                products: {
                   selectedProduct: mockProductCard
                }
            }

        });

        const productImage = getByAltText(mockProductCard.name);
        expect(productImage).toBeDefined();

        const productStatus = mockProductCard.status.charAt(0).toUpperCase() + mockProductCard.status.slice(1)
        const status = getByText(productStatus);
        expect(status).toBeDefined();

        const productName = getByText(mockProductCard.name);
        expect(productName).toBeDefined();

        const productCode = getByText(`Product code ${mockProductCard.id}`);
        expect(productCode).toBeDefined();

        const productPrice = getByText(`${mockProductCard.price.toFixed(2)}USD`);
        expect(productPrice).toBeDefined();

        const additionalInfo = getByText(mockProductCard.additional_info[0].value)
        expect(additionalInfo).toBeDefined();

        expect(asFragment()).toMatchSnapshot();
    })
})