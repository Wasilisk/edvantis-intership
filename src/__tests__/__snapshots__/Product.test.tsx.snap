// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product component should render header component 1`] = `
<DocumentFragment>
  <div
    class="sc-dkzDqf OyftB"
  >
    <div
      class="product-image"
    >
      <img
        alt="Lord armchair, Miedel Home"
        src="https://e-c.storage.googleapis.com/res/5f731c835a5a2667cb27290e/480"
      />
    </div>
    <div
      class="product-info"
    >
      <div
        class="text-info"
      >
        <p
          class="status highlight"
        >
          In stock
        </p>
        <p
          class="product-name"
        >
          Lord armchair, Miedel Home
        </p>
        <p
          class="product-code"
        >
          Product code 1
        </p>
        <p
          class="price"
        >
          120.50USD
        </p>
      </div>
      <div
        class="actions"
      >
        <div
          class="sc-gsnTZi ivbWhk"
        >
          <input
            data-testid="input"
            max="10"
            min="1"
            type="number"
            value="1"
          />
          <div
            class="input-arrows"
          >
            <svg
              data-testid="plus"
            >
              input-arrow.svg
            </svg>
            <svg
              class="down-arrow"
              data-testid="minus"
            >
              input-arrow.svg
            </svg>
          </div>
        </div>
        <button
          class="sc-eCYdqJ bWmvnW"
          type="button"
        >
          Add to card
        </button>
      </div>
      <div
        class="additional-info"
      >
        <div
          class="sc-hKMtZM hRjqbZ"
        >
          <div
            class="tabs"
            data-testid="tabs"
          >
            <div
              class="sc-bczRLJ cDnWoO"
            >
              <span>
                description
              </span>
            </div>
          </div>
        </div>
        <div
          class="tab-panel"
          data-testid="tab-panel-description"
          role="tabpanel"
        >
          <pre>
            description value
          </pre>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
